### PR TITLE
Add patch to fmt package

### PR DESCRIPTION
**Pull request by adayton1**
* Update to newer fmt package (make newer versions available)
* Add a patch for the clang12.0.1.ibm compiler

### CODE REVIEW

File: packages/fmt/package.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    """fmt (formerly cppformat) is an open-source formatting library.
+    It can be used as a safe alternative to printf or as a fast alternative
+    to C++ IOStreams."""
+
+    homepage = "https://fmt.dev/"
+    url = "https://github.com/fmtlib/fmt/releases/download/7.1.3/fmt-7.1.3.zip"
+    git = "https://github.com/fmtlib/fmt.git"
+    maintainers("msimberg")
+
+    version("10.2.1", sha256="312151a2d13c8327f5c9c586ac6cf7cddc1658e8f53edae0ec56509c8fa516c9")
```
Comment (adayton1) [inline packages/fmt/package.py:19]: I think the only differences in this file are the newer versions and the patches.

File: packages/fmt/package.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+    # Fix compilation with clang in CUDA mode: https://github.com/fmtlib/fmt/issues/3740
+    patch(
+        "https://github.com/fmtlib/fmt/commit/89860eb9013a345608c8144b1aad5f12b0682d7e.patch?full_index=1",
+        sha256="6ef12fe60a2b3625139c6d29c748dafd81b51e2a0690c1fa37604ed5b15615e0",
+        when="@10.0.0:10.1.1",
+    )
+
+    # Fix 'variable "buffer" may not be initialized' compiler error
+    patch("fmt-no-variable-initialize_10.0.0.patch", when="@10.0.0:10.2.1%clang@12.0.1.ibm.gcc.8.3.1")
```
Comment (adayton1) [inline packages/fmt/package.py:95]: Unfortunately, since this compiler is specific to us, I don't we'll be able to upstream this.

Comment (adayton1) [reply packages/fmt/package.py:95]: Or maybe they wouldn't care since it shouldn't do any harm.

**Comment by adrienbernede** [comment]
@adayton1 how critical is the need for those patches and new fmt version ?
Is this new version you want available in the latest spack release ? (i.e. spack@0.21.1).
I am trying not to add another local package if I can avoid it. Of course, I am willing to understand your need.

**Comment by adayton1** [comment]
> @adayton1 how critical is the need for those patches and new fmt version ? Is this new version you want available in the latest spack release ? (i.e. spack@0.21.1). I am trying not to add another local package if I can avoid it. Of course, I am willing to understand your need.

The newer versions found in spack@0.21.1 should be fine. However, the `fmt-no-variable-initialize_10.0.0.patch` file is necessary for the clang-ibm compilers (at least for clang@12.0.1.ibm). Everything else is just copied from spack@develop. Alternatively, we might be able to disable the build with that particular compiler, though I think some people may still be using it.

**Comment by kab163** [comment]
> > @adayton1 how critical is the need for those patches and new fmt version ? Is this new version you want available in the latest spack release ? (i.e. spack@0.21.1). I am trying not to add another local package if I can avoid it. Of course, I am willing to understand your need.
> 
> The newer versions found in spack@0.21.1 should be fine. However, the `fmt-no-variable-initialize_10.0.0.patch` file is necessary for the clang-ibm compilers (at least for [clang@12.0.1.ibm](mailto:clang@12.0.1.ibm)). Everything else is just copied from spack@develop. Alternatively, we might be able to disable the build with that particular compiler, though I think some people may still be using it.

@davidbeckingsale any thoughts on this? Clang@12.0.1 still needs to be in the CI, ya?

**Comment by adrienbernede** [comment]
@adayton1 thank you for your answer.

In https://github.com/LLNL/Umpire/pull/864 I'm  now using spack@0.21.1, but that won't cover for the needed patch.

I'll get your changes merged in radiuss spack configs tomorrow. I think it is a valid use of the local patching mechanism.

We can pursue pushing the patch upstream to Spack, it won't impact anyone, but since we need this working now, we will get it in radiuss spack configs first.

Note to self: tag current main if it _requires_ a newer spack than previous tag.

**Review by kab163 — APPROVED** [review]

**Comment by adrienbernede** [comment]
Thank you @adayton1 !
Please test that the resulting merge works for you and remove this branch once you have not use for it.

**Pull request merged**